pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,482 @@
+			Project:        a.projectID,
+		}
+		updated, err = a.globalForwardingRulesClient.Get(ctx, getReq)
+	} else {
+		getReq := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		updated, err = a.forwardingRulesClient.Get(ctx, getReq)
+	}
+	if err != nil {
+		return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+	}
+
+	// setTarget request is sent when there are updates to target.
+	if !reflect.DeepEqual(forwardingRule.Target, a.actual.Target) {
+		if a.Location == "global" {
+			setTargetReq := &computepb.SetTargetGlobalForwardingRuleRequest{
+				ForwardingRule:          a.resourceID,
+				TargetReferenceResource: &computepb.TargetReference{Target: forwardingRule.Target},
+				Project:                 a.projectID,
+			}
+			op, err = a.globalForwardingRulesClient.SetTarget(ctx, setTargetReq)
+		} else {
+			setTargetReq := &computepb.SetTargetForwardingRuleRequest{
+				ForwardingRule:          a.resourceID,
+				TargetReferenceResource: &computepb.TargetReference{Target: forwardingRule.Target},
+				Project:                 a.projectID,
+				Region:                  a.Location,
+			}
+			op, err = a.forwardingRulesClient.SetTarget(ctx, setTargetReq)
+		}
+		if err != nil {
+			return fmt.Errorf("updating ComputeForwardingRule target %s failed: %w", a.fullyQualifiedName(), err)
+		}
+		err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting ComputeForwardingRule %s update target failed: %w", a.fullyQualifiedName(), err)
+		}
+		log.V(2).Info("successfully updated ComputeForwardingRule target", "name", a.fullyQualifiedName())
+	}
+
+	// Get the updated resource
+	if a.Location == "global" {
+		getReq := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		updated, err = a.globalForwardingRulesClient.Get(ctx, getReq)
+	} else {
+		getReq := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		updated, err = a.forwardingRulesClient.Get(ctx, getReq)
+	}
+	if err != nil {
+		return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+	}
+
+	status := &krm.ComputeForwardingRuleStatus{
+		LabelFingerprint:  updated.LabelFingerprint,
+		CreationTimestamp: op.Proto().InsertTime,
+		SelfLink:          updated.SelfLink,
+	}
+	return setStatus(u, status)
+}
+
+func (a *forwardingRuleAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	// TODO(kcc)
+	return nil, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *forwardingRuleAdapter) Delete(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("deleting ComputeForwardingRule", "name", a.resourceID)
+
+	exist, err := a.Find(ctx)
```

Comment:
Have you tried deleting a non-exist forwarding rule? If the GCP service gives a clear signal. i.e. 404, then maybe we do not need this additional GCP call?

Reply:
Removed the `Find` call here and verified the http logs